Add heartbeat to test methods so that Travis CI does not abort latent tests.

F2006 opened the pull request:
Introduced a heartbeat message to the console output during unit tests and disabled the metrics put to the console from Titan.

Reference: https://github.com/awslabs/dynamodb-titan-storage-backend/issues/87

amcp commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+
+            this.configuredUnitTestName = unitTestName;
+            this.heartbeatTimer = new Timer("Unit test heartbeat timer");
+            this.heartbeatTimer.schedule(new TimerTask() {
```
pull the TimerTask anonymous inner class up to a regular class

F2006 replied:
Done

amcp commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+                    long currentRunTimeMilliseconds = System.currentTimeMillis() - heartbeatStartTime;
+
+                    if (intervals == configuredMaximumIntervals) {
+                        LOG.warn(String.format("Heartbeat - [%d] - %s - %dms.",
```
you should pass the LOG of TravisCiHeartbeat into the constructor of the TimerTask

amcp replied:
passing in as an argument will allow you to test the heartbeat with mockito and count the number of times you thumped.

F2006 replied:
Done

amcp commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+            LOG.info(String.format("Heartbeat - [started] - %s - 0ms",
+                    this.configuredUnitTestName));
+        } finally {
+            this.lock.unlock();
```
inside each travis CI container, tests run serially. Heartbeat is a quick thing so cant we just synchronize heartbeats on this class? after a jitter wait that is.

F2006 replied:
The heartbeat message itself it not inside a lock, the output is synchronised on the timer thread and there can only be one timer instance at a time.

amcp commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+ *
+ * @author Johan Jacobs
+ */
+public class TravisCiHeartbeat {
```
This heartbeat is very useful and could be used in another CI platform. Perhaps note your classes generality and maybe even add a TODO to rename to CiHeartbeat. Dont need to make this change in this CR, just saying we should make note of it.

F2006 replied:
I have renamed the class to CiHeartbeat.

amcp commented:
Also, I renamed this PR because now that I can read the diff, the only thing this PR does is add the heartbeat. As such, I think this PR should only contain one commit that contains all the changes you need for the heartbeat. So, when you force-push to your branch next time, please squash all your commits in this PR to one commit and then rebase off the awslabs:travis branch.

F2006 commented:
@amcp thanks for the feedback. :) Will get on to this quick.

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBGraphConcurrentTest.java:
```diff
@@ -41,9 +42,15 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBEventualGraphTest.java:
```diff
@@ -29,9 +34,14 @@
 */
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBEventualGraphTest.java:
```diff
@@ -29,9 +34,14 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/diskstorage/dynamodb/AbstractDynamoDBStoreTest.java:
```diff
@@ -52,9 +55,14 @@
 public abstract class AbstractDynamoDBStoreTest extends KeyColumnValueStoreTest
```
Done

F2006 commented on src/test/java/com/amazon/titan/diskstorage/dynamodb/AbstractDynamoDBStoreTest.java:
```diff
@@ -52,9 +55,14 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/diskstorage/dynamodb/AbstractDynamoDBMultiWriteStoreTest.java:
```diff
@@ -70,9 +73,14 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/diskstorage/dynamodb/AbstractDynamoDBLockStoreTest.java:
```diff
@@ -37,12 +41,17 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/diskstorage/dynamodb/AbstractDynamoDBIDAuthorityTest.java:
```diff
@@ -33,11 +37,17 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBGraphIterativeTest.java:
```diff
@@ -33,9 +38,14 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBGraphPerformanceMemoryTest.java:
```diff
@@ -29,9 +34,14 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBGraphSpeedTest.java:
```diff
@@ -24,20 +26,29 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBGraphTest.java:
```diff
@@ -37,7 +40,11 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBOLAPTest.java:
```diff
@@ -19,19 +19,29 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBOperationCountingTest.java:
```diff
@@ -29,9 +34,14 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/AbstractDynamoDBPartitionGraphTest.java:
```diff
@@ -31,9 +36,15 @@
+    @Rule
+    public TestName testName = new TestName();
+
+    private TravisCiHeartbeat travisCiHeartbeat;
```
Done

F2006 commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Done

F2006 commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+    private final ReentrantLock lock;
+    private boolean timerStarted;
+    private int intervals;
+    private long configuredHeartbeatInterval;
```
Done

F2006 commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+    private boolean timerStarted;
+    private int intervals;
+    private long configuredHeartbeatInterval;
+    private int configuredMaximumIntervals;
```
Done

F2006 commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+    private long configuredHeartbeatInterval;
+    private int configuredMaximumIntervals;
+    private String configuredUnitTestName;
+    private long heartbeatStartTime;
```
Done

F2006 commented on src/test/java/com/amazon/titan/graphdb/dynamodb/SingleDynamoDBGraphTest.java:
```diff
@@ -66,7 +66,7 @@ public SingleDynamoDBGraphTest()
     @Override
     public WriteConfiguration getConfiguration() {
         final WriteConfiguration wc = super.getConfiguration();
-        final String methodName = name.getMethodName();
```
The TestName rule is now in the AbstractDynamoDBGraphTest class, so no need to have the same rule overriding the base class in the implementing class.

F2006 commented on src/test/java/com/amazon/titan/testutils/TravisCiHeartbeat.java:
```diff
@@ -0,0 +1,126 @@
+
+    public void startHeartbeat(String unitTestName) {
+
+        this.lock.lock();
```
The lock was only for start and stop, catering for cases where it might be started/stopped outside of the JUnit lifecycles, but if the start/stop is used in the JUnit lifecycle (before and after) it is not needed at all, I have removed the lock.

F2006 commented:
I'm busy resolving the merge conflicts. Have to run to be at an event, will try to get it all sorted tonight still

amcp commented:
@F2006 Looking good so far. Thank you for adding tests, 
Can you squash d25fd2e488985255c4de486563bb6b8bbb4781b8 - 74b35a541c4a2b06df0f124d2e4b2d324f71b03a on your branch?
I am trying to bring your branch to a situation where it can be rebased and merged. For that to happen, the parent commit of your branch needs to be b09b07ab451d2222daaab62d48c2e6e17e5f4b34 (the HEAD of travis) and not e6435601107303cc9c33ade1a18678da05aa1183 (the HEAD of 1.0.0).

amcp commented:
<img width="728" alt="rebase" src="https://cloud.githubusercontent.com/assets/1066771/25539924/2cbcead8-2c84-11e7-84e4-d345eec5247b.png">
Here is a visual from the network diagram.

amcp commented:
@F2006 The fewer commits you have, the easier it will be to rebase. That is why I suggest squashing the intermediate ones.

amcp commented:
Great, you got rid of the merge commit. now you can squash everything down to one commit and then rebase on upstream/travis!

F2006 commented:
@amcp I think it should be all good now. Definitely upped my git skill levels with a few points. :) 

I have addresses all the feedback above. If you can just give the changes a once over again it would be great, just incase I have missed something while fixing merge conflicts.
